spec/acceptance/pip_spec.rb

```diff
@@ -71,4 +70,36 @@ class { 'python':
     its(:exit_status) { is_expected.to eq 0 }
     its(:stdout) { is_expected.not_to match %r{agent.* 0\.1\.2} }
   end
+
+  context 'fails to install package with wrong version' do
+    it 'works with no errors' do
+      pp = <<-PUPPET
+      class { 'python':
+        version => '3',
+        dev     => 'present',
+      }
+
+      python::pyvenv { '/opt/test-venv':
+        ensure      => 'present',
+        systempkgs  => false,
+        mode        => '0755',
+        pip_version => '<= 20.3.4',
+      }
+
+      python::pip { 'agent package':
+        virtualenv => '/opt/test-venv',
+        pkgname    => 'agent',
+        ensure     => '0.1.33+2020-this_is_something-fun',
+      }
+      PUPPET
+
+      apply_manifest(pp, catch_failures: false)
+      apply_manifest(pp, catch_changes: true)
+    end
+  end
+
+  describe command('/opt/test-venv/bin/pip show agent') do
+    its(:exit_status) { is_expected.to eq 1 }
+    its(:stdout) { is_expected.not_to match %r{WARNING: Package\(s\) not found: agent} }
+  end
```

Comment:
Testing that something does not work in acceptance tests is a bit awkward because you have to test for the very exact error you are expecting.  Here, Puppet cannot apply the catalog (because no package exist with this name and version) so the test should ensure that this is the exact error we are seeing.

It is probably easier to do a unit test in `spec/defines/pip_spec.rb`: setup `python::pip` with your parameters and ensure that the expected resources are declared with the expected parameters.

Reply:
@smortex I understand, but there is a problem. [Indico](https://indico.cern.ch/) is an application made by CERN, but we are customizing it, and those version numbers come from a colleague of mines. Unfortunately, they're not available in a public repository and they can't be used for this test. 
On another note, this is the first time that I see "Github Actions", and the problem is that it's set to run on PRs, and I can't seem to find a way to run the test on my forked repo. I'll investigate a bit more.

Reply:
now it's gonna run. It will prove that the package manager is capable of handling a failure on its own.